pkg/registry/input_stream.go

```diff
@@ -0,0 +1,209 @@
+package registry
+
+import (
+	"fmt"
+
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+)
+
+type ReplacesInputStream struct {
+	graph    GraphLoader
+	packages map[string]map[string]*ImageInput
+}
+
+func NewReplacesInputStream(graph GraphLoader, toAdd []*ImageInput) (*ReplacesInputStream, error) {
+	stream := &ReplacesInputStream{
+		graph:    graph,
+		packages: map[string]map[string]*ImageInput{},
+	}
+
+	// Sort the bundle images into buckets by package
+	for _, image := range toAdd {
+		pkg := image.Bundle.Package
+		if _, ok := stream.packages[pkg]; !ok {
+			stream.packages[pkg] = map[string]*ImageInput{}
+		}
+		stream.packages[pkg][image.Bundle.Name] = image
+	}
+
+	// Validate each package, dropping any invalid packages so that the stream can still be used with remaining packages.
+	var errs []error
+	for pkg, images := range stream.packages {
+		// Multiple input images require some ordering between them -- using skips and replaces -- to ensure an index add operation is deterministic
+		connected, err := graphConnected(images)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("Error determining connectedness of package graph %s: %s", pkg, err))
```

Comment:
Isn't there slightly more to this? Don't we need to divide them up into a graph per channel? You could, presumably, add two bundles that aren't connected at all. In that case, you could add them in any order and it shouldn't return an error.

Reply:
> Don't we need to divide them up into a graph per channel?

I don't think we _need_ to. I imposed strict rules to simplify our logic and the edge cases we need to deal with. How often will we be adding bundles to a package with diverging branches?

Reply:
I think it's actually a fairly common case at the moment: you build release channels based on minor or major versions, then use a skiprange in the higher channel to determine what set of things can be upgraded from one to the next.

Reply:
you're right, I'll need to rethink this.

Reply:
Ah, I think the key here is that if either bundle defines a different default channel, then the order does matter.